File: apps/apple/core/Sources/LorvexDomain/HlcState.swift
import Foundation

/// Mutable HLC generator state for a single device, holding the
/// monotonicity invariant: every generated HLC is strictly greater than
/// the previous, even if the physical clock goes backward.
///
/// Surfaces hold one instance behind their own lock (the type is `final class`
/// so all references see the same counter advance).
public final class HlcState {
  public private(set) var lastPhysicalMs: UInt64
  public private(set) var counter: UInt32
  public let deviceSuffix: String

  /// Construct a generator. The suffix is validated up front (via
  /// `Hlc(physicalMs:counter:deviceSuffix:)`) and canonicalized to
  /// lowercase so every subsequent `generate(...)` is infallible.
  public init(deviceSuffix: String) throws {
    let canon = try Hlc(physicalMs: 0, counter: 0, deviceSuffix: deviceSuffix)
    self.lastPhysicalMs = 0
    self.counter = 0
    self.deviceSuffix = canon.deviceSuffix
  }

  /// Generate an HLC using the supplied physical timestamp (Unix ms).
  /// Deterministic seam for testing — production callers use `generate()`.
  ///
  /// Clamps inputs and post-bump physical to `Hlc.maxPhysicalMs` so a
  /// future-dated NTP response or counter-overflow tail can't poison the
  /// lex-sort ceiling.
  @discardableResult
  public func generate(withPhysicalMs physicalMs: UInt64) -> Hlc {
    let phys = min(physicalMs, Hlc.maxPhysicalMs)
    lastPhysicalMs = min(lastPhysicalMs, Hlc.maxPhysicalMs)
    let newPhysical = max(lastPhysicalMs, phys)

    if newPhysical == lastPhysicalMs {
      // Same or backward clock — increment counter (saturating against u32
      // overflow from a tainted prior value; the MAX_COUNTER guard below
      // rolls forward immediately).
      counter = counter == UInt32.max ? UInt32.max : counter &+ 1
    } else {
      counter = 0
    }
    lastPhysicalMs = newPhysical

    if counter > Hlc.maxCounter {
      // Overflow recovery: advance physical by 1 ms and reset the counter. At the
      // absolute lex-sort ceiling physical cannot advance, so SATURATE the counter
      // at its max instead of resetting to 0 — resetting there would emit
      // `(maxPhysicalMs, 0)` right after `(maxPhysicalMs, maxCounter)`, a
      // backwards-sorting HLC that corrupts LWW ordering. Holding at
      // `(maxPhysicalMs, maxCounter)` keeps the stream monotonic non-decreasing
      // (rare ties resolve via the existing tie-break). Only reachable at the
      // far-future ceiling year or via a peer HLC pinned to maxPhysicalMs.
      if lastPhysicalMs == Hlc.maxPhysicalMs {
        counter = Hlc.maxCounter
      } else {
        lastPhysicalMs += 1
        counter = 0
      }
    }

    // Construction can only fail on suffix-shape violations, which are
    // ruled out by the constructor's up-front check.
    return Hlc(uncheckedPhysicalMs: lastPhysicalMs, counter: counter, deviceSuffix: deviceSuffix)
  }

  /// Generate using the current wall-clock time. Falls back to 0 on a
  /// pre-1970 clock so HLC generation never crashes the writer.
  @discardableResult
  public func generate() -> Hlc {
    let now = Date().timeIntervalSince1970
    let nowMs: UInt64 = now < 0 ? 0 : UInt64(now * 1000)
    return generate(withPhysicalMs: nowMs)
  }

  /// Default forward clock-skew bound for an inbound PEER envelope's HLC (ms).
  /// A peer with a wrong/future clock must not be able to drag this device's
  /// clock arbitrarily far forward — that would pin the whole fleet's clock and
  /// let the bad envelope's future timestamp dominate every honest concurrent
  /// edit. Five minutes dwarfs honest NTP skew (sub-second) yet sits far below
  /// `Hlc.maxPhysicalMs`, so within-bound the clamp is a literal no-op.
  public static let maxInboundForwardDriftMs: UInt64 = 5 * 60 * 1000

  /// Advance local state after receiving a remote HLC, so the next
  /// locally-minted HLC is strictly greater than the remote one. `physicalMs` is
  /// the current wall-clock at the moment of receive.
  ///
  /// UNBOUNDED — use only for advancing past a future row `version` an EXPLICIT
  /// local edit must supersede: a deliberate user mutation must always be able to
  /// dominate the row it edits. Every OBSERVED value whose physical time is not
  /// trusted — an inbound PEER envelope, the startup seed scan (rows matched by a
  /// device suffix a peer could forge), and a merge/clear HLC minted during apply
  /// from attacker-influenced participant versions — uses the bounded overload
  /// ``updateOnReceive(remote:physicalMs:maxForwardDriftMs:)`` instead, so a
  /// future-clocked value cannot pin this device's clock.
  public func updateOnReceive(remote: Hlc, physicalMs: UInt64) {
    applyRemote(
      remotePhysical: remote.physicalMs, remoteCounter: remote.counter, physicalMs: physicalMs)
  }

  /// Advance local state after OBSERVING an HLC whose physical time is not
  /// trusted — an inbound PEER envelope, the startup seed scan (rows matched by a
  /// forgeable device suffix), or a merge/clear HLC minted during apply from
  /// attacker-influenced participant versions — bounding the remote's forward
  /// reach to `phys + maxForwardDriftMs`. Forward-only: a past-dated remote is
  /// untouched. This is per-device clock hygiene only — it never changes which
  /// version wins LWW (a pure lexicographic compare of version strings on every
  /// peer), so a mixed fleet (some devices bounding, some not) still converges;
  /// only the local clock advance is contained, so no single untrusted future
  /// stamp — arriving on an envelope OR driving a merge HLC — can pin this
  /// device's clock beyond `now + drift`. The lone exception is an explicit local
  /// edit, which advances unbounded via ``updateOnReceive(remote:physicalMs:)`` so
  /// the user can always supersede the row they are editing.
  public func updateOnReceive(remote: Hlc, physicalMs: UInt64, maxForwardDriftMs: UInt64) {
    let phys = min(physicalMs, Hlc.maxPhysicalMs)
    let cappedRemotePhys = min(remote.physicalMs, phys &+ maxForwardDriftMs)
    applyRemote(
      remotePhysical: cappedRemotePhys, remoteCounter: remote.counter, physicalMs: physicalMs)
  }

  private func applyRemote(remotePhysical: UInt64, remoteCounter: UInt32, physicalMs: UInt64) {
    let phys = min(physicalMs, Hlc.maxPhysicalMs)
    lastPhysicalMs = min(lastPhysicalMs, Hlc.maxPhysicalMs)
    let remotePhys = min(remotePhysical, Hlc.maxPhysicalMs)
    let newPhysical = max(max(lastPhysicalMs, remotePhys), phys)

    if newPhysical == lastPhysicalMs && newPhysical == remotePhys {
      let m = max(counter, remoteCounter)
      counter = m == UInt32.max ? UInt32.max : m &+ 1
    } else if newPhysical == lastPhysicalMs {
      counter = counter == UInt32.max ? UInt32.max : counter &+ 1
    } else if newPhysical == remotePhys {
      counter = remoteCounter == UInt32.max ? UInt32.max : remoteCounter &+ 1
    } else {
      counter = 0
    }
    lastPhysicalMs = newPhysical

    if counter > Hlc.maxCounter {
      // See `generate(withPhysicalMs:)`: saturate at the absolute ceiling rather
      // than wrapping the counter backwards.
      if lastPhysicalMs == Hlc.maxPhysicalMs {
        counter = Hlc.maxCounter
      } else {
        lastPhysicalMs += 1
        counter = 0
      }
    }
  }
}

/// Storage-side handle that backs an `HlcSession`. Each surface
/// implements this for whatever container holds its `HlcState`.
public protocol HlcStateHandle: Sendable {
  /// Acquire the storage lock and mint the next strictly-monotonic HLC.
  /// Implementations must serialize concurrent calls.
  func generate() -> Hlc
}

/// Optional stronger clock capability used by explicit local operations that
/// must supersede a known future-stamped row. Keeping it separate from
/// ``HlcStateHandle`` preserves lightweight deterministic test handles while a
/// production clock can provide an atomic advance-and-mint implementation.
public protocol HlcDominatingStateHandle: HlcStateHandle {
  func generate(dominating floor: Hlc?) -> Hlc
}

/// Per-mutation HLC stamp session. Constructed once per top-level
/// mutation; `nextVersion()` is called once per row stamped within that
/// mutation. Unifies the API boundary across surfaces without touching
/// how each surface stores its `HlcState`.
public struct HlcSession {
  private let handle: any HlcStateHandle

  public init(handle: any HlcStateHandle) { self.handle = handle }

  /// Mint the next strictly-monotonic HLC.
  public func nextVersion() -> Hlc { handle.generate() }

  /// Mint after `floor` when the backing clock supports explicit dominance.
  /// Lightweight test handles fall back to their ordinary monotonic mint; the
  /// caller that requires dominance still validates the returned HLC.
  public func nextVersion(dominating floor: Hlc?) -> Hlc {
    if let dominating = handle as? any HlcDominatingStateHandle {
      return dominating.generate(dominating: floor)
    }
    return handle.generate()
  }

  /// Convenience: mint the next stamp and stringify it to the wire
  /// format. Avoids a `.description` at every call site.
  public func nextVersionString() -> String { nextVersion().description }

  public func nextVersionString(dominating floor: Hlc?) -> String {
    nextVersion(dominating: floor).description
  }
}

File: apps/tauri/lorvex-domain/src/hlc_state.rs
//! Mutable HLC generator state.
//!
//! Maintains the monotonicity invariant: every generated HLC is strictly
//! greater than the previous, even if the physical clock goes backward.
//!
//! # Key invariant
//!
//! `generate_with_physical` always sets `last_physical_ms = max(last_physical_ms,
//! now)` before branching on whether to increment the counter or reset it.

use crate::hlc::{Hlc, HlcParseError, MAX_HLC_PHYSICAL_MS};

/// Maximum counter value. On overflow (>9999 writes in the same ms on one
/// device), recovery advances physical_ms by 1 and resets counter to 0.
pub use crate::hlc::MAX_COUNTER;

/// Mutable state for HLC generation on a single device.
///
/// Fields are `pub(crate)` so external callers must go through the
/// `last_physical_ms()` / `counter()` / `device_suffix()` accessors
/// instead of poking the fields directly.
/// at the bottom of `hlc.rs` already documented that direct field
/// mutation would corrupt the LWW invariant; tightening visibility
/// makes the rule structurally enforceable rather than just
/// documented. The in-crate proptest / overflow tests can still seed
/// adversarial state because they live inside `lorvex-domain`.
#[derive(Debug)]
pub struct HlcState {
    pub(crate) last_physical_ms: u64,
    pub(crate) counter: u32,
    pub(crate) device_suffix: String,
}

impl HlcState {
    /// Read-only access to the current state for diagnostic /
    /// assertion sites (e.g. `lorvex-store`'s seeding tests).
    #[inline]
    pub const fn last_physical_ms(&self) -> u64 {
        self.last_physical_ms
    }

    /// Read-only access to the in-millisecond counter.
    #[inline]
    pub const fn counter(&self) -> u32 {
        self.counter
    }

    /// Read-only access to the canonical device suffix.
    #[inline]
    pub fn device_suffix(&self) -> &str {
        &self.device_suffix
    }

    /// Create a new HLC generator state for the given device.
    ///
    /// the suffix is validated up front by minting a
    /// throwaway `Hlc` so an invalid shape (wrong length / non-hex
    /// alphabet) surfaces immediately rather than at the first
    /// `generate_with_physical` call. The constructed value also
    /// canonicalizes the suffix to lowercase so subsequent
    /// `Hlc::new` calls inside `generate_with_physical` stay
    /// infallible.
    pub fn new(device_suffix: impl Into<String>) -> Result<Self, HlcParseError> {
        let suffix_str = device_suffix.into();
        // Validation lives in `Hlc::new`; reuse it so a single source
        // of truth governs what counts as a canonical suffix.
        let canon = Hlc::new(0, 0, &suffix_str)?;
        Ok(Self {
            last_physical_ms: 0,
            counter: 0,
            device_suffix: canon.device_suffix().to_string(),
        })
    }

    /// Generate an HLC using the provided physical timestamp (milliseconds).
    ///
    /// This is the core generation function. Use `generate()` for wall-clock
    /// time, or call this directly with a known timestamp for deterministic
    /// testing.
    pub fn generate_with_physical(&mut self, physical_ms: u64) -> Hlc {
        // clamp to the 13-digit lex-sortable ceiling before
        // it propagates. A future-dated NTP response or deliberately
        // mis-set local clock could otherwise emit a version that lex-
        // sorts above every legitimate entry forever, and LWW would
        // promote that poisoned value cluster-wide. Also sanitize the
        // stored `last_physical_ms` in case a tainted on-disk history
        // (or test harness) seeded it past the ceiling — `Hlc::new`
        // would otherwise panic at the bottom of this function.
        let physical_ms = physical_ms.min(MAX_HLC_PHYSICAL_MS);
        self.last_physical_ms = self.last_physical_ms.min(MAX_HLC_PHYSICAL_MS);
        let new_physical = std::cmp::max(self.last_physical_ms, physical_ms);

        if new_physical == self.last_physical_ms {
            // Same or backward clock: increment counter.
            // saturating_add protects against u32 overflow if a prior
            // malicious / malformed peer envelope forced the counter
            // to u32::MAX. The standard MAX_COUNTER guard below will
            // roll the clock forward immediately after.
            self.counter = self.counter.saturating_add(1);
        } else {
            // Clock advanced: reset counter.
            self.counter = 0;
        }

        self.last_physical_ms = new_physical;

        // Counter overflow recovery: advance physical time by 1ms, reset counter.
        // This is the standard HLC recovery for degenerate scenarios with >9999
        // writes in the same millisecond (e.g. batch import tight loops).
        //
        // clamp the post-bump physical to `MAX_HLC_PHYSICAL_MS`.
        // Without this clamp, the pathological case `last_physical_ms ==
        // MAX_HLC_PHYSICAL_MS && counter == MAX_COUNTER` bumps the stored ms
        // to `MAX_HLC_PHYSICAL_MS + 1`, escaping the 13-digit lex-sort
        // ceiling by one millisecond and breaking comparison against any
        // HLC written by a pre-overflow device. Saturating at the ceiling
        // keeps the same millisecond and just holds the counter at 0 —
        // liveness preserved, lex-sort invariant preserved.
        //
        // the race-audit explicitly verified
        // this branch as informational/safe — the saturating-add chain
        // (`#2233` on counter, `#2741` on physical_ms) plus the ceiling
        // clamp here means there is no path that emits an HLC larger
        // than the lex-sort ceiling, regardless of how malformed an
        // inbound peer envelope is or how tight a local loop pushes the
        // counter. The proptest below (`generate_saturating_on_max_u32_local_counter`)
        // pins the contract; do not weaken either guard without
        // updating that test.
        if self.counter > MAX_COUNTER {
            self.last_physical_ms = self
                .last_physical_ms
                .saturating_add(1)
                .min(MAX_HLC_PHYSICAL_MS);
            self.counter = 0;
        }

        // The suffix invariant is enforced at `HlcState::new`; if a
        // caller mutated `self.device_suffix` to something invalid we
        // surface that as a panic here — better to crash the writer
        // than to emit a malformed HLC the cluster will accept.
        Hlc::new(self.last_physical_ms, self.counter, &self.device_suffix)
            .expect("HlcState.device_suffix invariant violated after construction")
    }

    /// Generate an HLC using the current wall-clock time.
    pub fn generate(&mut self) -> Hlc {
        // `duration_since(UNIX_EPOCH)` returns Err on a pre-1970
        // clock (devices without an RTC battery, VMs restored from
        // bad snapshots, deliberately-mis-set test clocks).
        // `.expect()`-ing this would panic on every HLC generation
        // and crash the Tauri app, the MCP server, and the CLI.
        // Fall back to Duration::ZERO and let
        // `generate_with_physical` preserve monotonicity via the
        // counter — HLC is explicitly designed for this case.
        let now_ms = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or(std::time::Duration::ZERO)
            .as_millis() as u64;
        self.generate_with_physical(now_ms)
    }

    /// Update local state after receiving a remote HLC.
    ///
    /// Ensures subsequent local HLCs are strictly greater than the remote one.
    /// `physical_ms` is the current wall-clock time at the moment of receive.
    pub fn update_on_receive(&mut self, remote: &Hlc, physical_ms: u64) {
        // clamp the local wall clock to the accepted range
        // so a future-dated NTP response cannot push state past the
        // 13-digit lex-sort ceiling. The matching
        // `remote.physical_ms.min(MAX_HLC_PHYSICAL_MS)` clamp was
        // removed because every `Hlc` value is now guaranteed by the
        // type's constructors (`Hlc::new` and `Hlc::parse`) to satisfy
        // `physical_ms <= MAX_HLC_PHYSICAL_MS` — the defense-in-depth
        // clamp that live here can no longer observe a
        // violating value.
        let physical_ms = physical_ms.min(MAX_HLC_PHYSICAL_MS);
        // Sanitize a seeded-past-ceiling local state for the same
        // reason as `generate_with_physical` — a tainted on-disk
        // history would otherwise survive every receive and keep
        // emitting >13-digit physicals after the next generate.
        self.last_physical_ms = self.last_physical_ms.min(MAX_HLC_PHYSICAL_MS);
        let remote_physical_ms = remote.physical_ms();
        let new_physical = std::cmp::max(
            std::cmp::max(self.last_physical_ms, remote_physical_ms),
            physical_ms,
        );

        // `remote.counter` comes from a peer-supplied
        // HLC string parsed as u32, so the `+ 1` below could panic
        // (debug) or wrap to 0 (release) on a malicious / malformed
        // envelope with counter ≈ u32::MAX. Use saturating_add so the
        // MAX_COUNTER recovery guard below catches it and rolls the
        // physical clock forward, preserving monotonicity.
        if new_physical == self.last_physical_ms && new_physical == remote_physical_ms {
            // All three timestamps are the same: take max counter + 1.
            self.counter = std::cmp::max(self.counter, remote.counter()).saturating_add(1);
        } else if new_physical == self.last_physical_ms {
            // Local is the max: increment local counter.
            self.counter = self.counter.saturating_add(1);
        } else if new_physical == remote_physical_ms {
            // Remote is the max: continue from remote counter + 1.
            self.counter = remote.counter().saturating_add(1);
        } else {
            // Wall clock is strictly ahead of both: reset counter.
            self.counter = 0;
        }

        self.last_physical_ms = new_physical;

        // Counter overflow recovery: advance physical time by 1ms,
        // reset counter. Clamp the bump
        // to `MAX_HLC_PHYSICAL_MS` to match the generate path.
        // Without the clamp, a pathological case where both sides
        // are at the ceiling and the counter saturates would push
        // `last_physical_ms` to `MAX_HLC_PHYSICAL_MS + 1` — the
        // resulting HLC string would be 14 digits in the physical
        // half and lex-sort above every legitimate 13-digit entry,
        // poisoning every LWW comparison from this device forward.
        // Saturating at the ceiling preserves the lex-sort invariant
        // at the cost of holding the counter at 0 in this same ms,
        // which is the standard hold-and-retry HLC degenerate case.
        if self.counter > MAX_COUNTER {
            self.last_physical_ms = self
                .last_physical_ms
                .saturating_add(1)
                .min(MAX_HLC_PHYSICAL_MS);
            self.counter = 0;
        }
    }
}

#[cfg(test)]
mod tests;
